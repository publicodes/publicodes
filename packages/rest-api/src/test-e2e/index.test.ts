import Router from '@koa/router'
import chai from 'chai'
import chaiHttp from 'chai-http'
import Koa from 'koa'
import Engine from 'publicodes'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import { parse } from 'yaml'
import publicodesAPI from '../middleware/koa'

interface State extends Koa.DefaultState {}

interface Context extends Koa.DefaultContext {}

const app = new Koa<State, Context>()
const router = new Router<State, Context>()

const apiRoutes = publicodesAPI(
	new Engine(
		parse(`
coucou: 0
coucou . j'ai des caractères spéciaux: "'ok'"
`),
	),
)

router.use(apiRoutes)

app.use(router.routes())
app.use(router.allowedMethods())

const server = app.listen(3003)

chai.use(chaiHttp)

beforeAll(async () => {
	await new Promise((res, _rej) => {
		server.once('listening', () => res(null))
	})
})

afterAll(async () => {
	server.close()
})

describe('e2e koa middleware', () => {
	it('Test evaluate endpoint', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({ expressions: '1 + 1' })
				.then(async (res) => {
					expect(res.status).toMatchInlineSnapshot('200')

					return JSON.parse(res.text)
				}),
		).resolves.toMatchInlineSnapshot(
			`
			{
			  "evaluate": [
			    {
			      "missingVariables": {},
			      "nodeValue": 2,
			    },
			  ],
			  "warnings": [],
			}
		`,
		)
	})

	it('Test evaluate endpoint with bad expression', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({ expressions: '1+1' })
				.then(async (res) => {
					expect(res.status).toMatchInlineSnapshot('200')

					return JSON.parse(res.text)
				}),
		).resolves.toMatchSnapshot()
	})

	it('Test evaluate endpoint with null expression', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({ expressions: null })
				.then(async (res) => {
					expect(res.status).toMatchInlineSnapshot('400')

					return JSON.parse(res.text)
				}),
		).resolves.toMatchSnapshot()
	})

	it('Test evaluate endpoint with empty expression', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({ expressions: [] })
				.then(async (res) => {
					expect(res.status).toMatchInlineSnapshot('400')

					return JSON.parse(res.text)
				}),
		).resolves.toMatchSnapshot()
	})

	it('Test evaluate endpoint with bad situation', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({
					expressions: ['coucou = 2 * 21'],
					situation: 'coucou: 42',
				})
				.then(async (res) => {
					expect(res.status).toMatchInlineSnapshot('400')

					return JSON.parse(res.text)
				}),
		).resolves.toMatchInlineSnapshot(`
			[
			  {
			    "dataPath": ".situation",
			    "keyword": "type",
			    "message": "should be object",
			    "params": {
			      "type": "object",
			    },
			    "schemaPath": "#/properties/situation/type",
			  },
			]
		`)
	})

	it('Test evaluate endpoint with situation', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({
					expressions: ['coucou = 2 * 21'],
					situation: { coucou: '42' },
				})
				.then(async (res) => {
					expect(res.status).toMatchInlineSnapshot('200')

					return JSON.parse(res.text)
				}),
		).resolves.toMatchInlineSnapshot(`
			{
			  "evaluate": [
			    {
			      "missingVariables": {},
			      "nodeValue": true,
			    },
			  ],
			  "warnings": [],
			}
		`)
	})

	it('Test rules endpoint', async () => {
		await expect(
			chai
				.request(server)
				.get('/rules')
				.then(async (res) => {
					expect(res.status).toMatchInlineSnapshot('200')

					return JSON.parse(res.text)
				}),
		).resolves.toMatchInlineSnapshot(`
			{
			  "coucou": {
			    "nodeKind": "rule",
			    "rawNode": {
			      "valeur": "0",
			    },
			    "replacements": [],
			    "suggestions": {},
			    "title": "Coucou",
			  },
			  "coucou . j'ai des caractères spéciaux": {
			    "nodeKind": "rule",
			    "rawNode": {
			      "valeur": "'ok'",
			    },
			    "replacements": [],
			    "suggestions": {},
			    "title": "J'ai des caractères spéciaux",
			  },
			}
		`)
	})

	it('Test rules/:id endpoint', async () => {
		await expect(
			chai
				.request(server)
				.get('/rules/coucou')
				.then(async (res) => {
					expect(res.status).toMatchInlineSnapshot('200')

					return JSON.parse(res.text)
				}),
		).resolves.toMatchInlineSnapshot(`
			{
			  "nodeKind": "rule",
			  "rawNode": {
			    "valeur": "0",
			  },
			  "replacements": [],
			  "suggestions": {},
			  "title": "Coucou",
			}
		`)
	})

	it('Test rules/:id endpoint with special characters', async () => {
		await expect(
			chai
				.request(server)
				.get(encodeURI("/rules/coucou . j'ai des caractères spéciaux"))
				.then(async (res) => {
					expect(res.status).toMatchInlineSnapshot('200')
					return JSON.parse(res.text)
				}),
		).resolves.toMatchInlineSnapshot(`
			{
			  "nodeKind": "rule",
			  "rawNode": {
			    "valeur": "'ok'",
			  },
			  "replacements": [],
			  "suggestions": {},
			  "title": "J'ai des caractères spéciaux",
			}
		`)
	})

	it('Test bad rules/:id endpoint', async () => {
		await expect(
			chai
				.request(server)
				.get('/rules/bad rule')
				.then(async (res) => {
					expect(res.status).toMatchInlineSnapshot('200')

					return JSON.parse(res.text)
				}),
		).resolves.toMatchInlineSnapshot(`
			{
			  "error": {
			    "message": "
			[ Règle inconnue ]
			➡️  Dans la règle \\"bad rule\\"
			⚠️  La règle 'bad rule' n'existe pas",
			  },
			}
		`)
	})
})
