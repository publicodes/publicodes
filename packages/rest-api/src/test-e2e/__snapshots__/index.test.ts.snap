// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`e2e koa middleware > Test evaluate endpoint with bad expression 2`] = `
{
  "evaluate": [
    {
      "error": {
        "message": "
[ Erreur syntaxique ]
➡️  Dans la règle \\"$EVALUATION\\"
✖️  L'expression suivante n'est pas valide :
   
   1+1
    ^
   Les opérateurs doivent être entourés d’espaces (\\"2 + 2\\" et non \\"2+2\\")
",
      },
      "warnings": [],
    },
  ],
  "warnings": [],
}
`;

exports[`e2e koa middleware > Test evaluate endpoint with empty expression 2`] = `
[
  {
    "dataPath": ".expressions",
    "keyword": "minItems",
    "message": "should NOT have fewer than 1 items",
    "params": {
      "limit": 1,
    },
    "schemaPath": "#/properties/expressions/oneOf/0/minItems",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be number",
    "params": {
      "type": "number",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/0/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be string",
    "params": {
      "type": "string",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/1/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be object",
    "params": {
      "type": "object",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/2/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "oneOf",
    "message": "should match exactly one schema in oneOf",
    "params": {
      "passingSchemas": null,
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf",
  },
  {
    "dataPath": ".expressions",
    "keyword": "oneOf",
    "message": "should match exactly one schema in oneOf",
    "params": {
      "passingSchemas": null,
    },
    "schemaPath": "#/properties/expressions/oneOf",
  },
]
`;

exports[`e2e koa middleware > Test evaluate endpoint with null expression 2`] = `
[
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be array",
    "params": {
      "type": "array",
    },
    "schemaPath": "#/properties/expressions/oneOf/0/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be number",
    "params": {
      "type": "number",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/0/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be string",
    "params": {
      "type": "string",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/1/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be object",
    "params": {
      "type": "object",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/2/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "oneOf",
    "message": "should match exactly one schema in oneOf",
    "params": {
      "passingSchemas": null,
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf",
  },
  {
    "dataPath": ".expressions",
    "keyword": "oneOf",
    "message": "should match exactly one schema in oneOf",
    "params": {
      "passingSchemas": null,
    },
    "schemaPath": "#/properties/expressions/oneOf",
  },
]
`;
